### ISSUE TITLE
Conditional and types ?

### ISSUE THREAD

**Issue by MarcWeber** [vimeo/psalm#10604]
If you have code like this:

```php
require_once "foo".$x.'.php';

if ($x == "a"){
  echo (new X())->a();
}
```

Assuming foo* files load class X whose methods depend on the file being loaded.
Same about 

```php

$row = query( "SELECT ." .($admin ? "field_for_admin," : "")." FROM table" );

if ($admin)
   echo $row['field_for-admin']
```

like code. So it would require hint such as "follow all possible branches/values of $admin which can be true/false and test each of them differently". ?

Does this even make sense ? It's important to understand that the admin case might fire some longer running SQL queries thus you cannot just get all data always without compromising performance.

**Comment by MarcWeber** [comment]
The idea was to see the same code with 2 eyes: If user is admin, if user is a user. If its admin some rows (eg from database) might have additional fields. like this:

```php
<?php

$admin = $_SESSION['admin'];

$rows = sql('SELECT name '.($admin? ', hidden_field' : '').' FROM table ');
# from now on rows either has rows with name or with name and hidden_field


echo $rows[0]['name']; // ok
echo $rows[0]['hidden_field']; // maybe not ok cause could be null but probably you want this to be an error

if ($admin){ 
   echo $rows[0]['hidden_field']; // definitely ok
}

// so basically rows has a type such as if $admin evauates to boolean true then array{name: string, hidden_field: srting} else array{name:string}
```

**Comment by kkmuffme** [comment]
This is already possible with @psalm-assert-if-... respectively this-out https://psalm.dev/docs/annotating_code/assertion_syntax/ isn't it?

It requires an extra function instead of just `if ($admin){` it's quite easily doable with what's available and I'm not sure how this could be made significantly easier. In your example how do you know that in case of $admin it will have hidden_field in the returned array? That's not declared anywhere in the code, it's just what you claim - if it's inferable automatically based on the code psalm does/would already do it anyway.

**Comment by MarcWeber** [comment]
You're right that it would require analysing SQL in that example. Closing for now.
